python/ray/tune/experimental/output.py

```diff
@@ -621,13 +647,109 @@ def _time_heartbeat_str(self):
+        if force or time.time() - self._last_heartbeat_time >= self._heartbeat_freq:
             self._print_heartbeat(trials, *args, force=force)
             self._last_heartbeat_time = time.time()
 
     def _print_heartbeat(self, trials, *args, force: bool = False):
         raise NotImplementedError
 
+    def _print_result(self, trial, result: Optional[Dict] = None, force: bool = False):
+        """Only print result if a different result has been reported, or force=True"""
+        result = result or trial.last_result
+
+        last_result_iter = self._trial_last_printed_results.get(trial.trial_id, -1)
+        this_iter = result.get(TRAINING_ITERATION, 0)
+
+        if this_iter != last_result_iter or force:
+            _print_dict_as_table(
+                result,
+                header=f"{self._addressing_tmpl.format(trial)} result",
+                include=self._progress_metrics,
+                exclude=BLACKLISTED_KEYS,
+                division=AUTO_RESULT_KEYS,
+            )
+            self._trial_last_printed_results[trial.trial_id] = this_iter
+
+    def _print_config(self, trial):
+        _print_dict_as_table(
+            trial.config, header=f"{self._addressing_tmpl.format(trial)} config"
+        )
+
+    def on_trial_result(
+        self,
+        iteration: int,
+        trials: List[Trial],
+        trial: Trial,
+        result: Dict,
+        **info,
+    ):
+        if self.verbosity < self._intermediate_result_verbosity:
+            return
+        curr_time_str, running_time_str = _get_time_str(self._start_time, time.time())
+        print(
+            f"{self._addressing_tmpl.format(trial)} "
+            f"finished iteration {result[TRAINING_ITERATION]} "
+            f"at {curr_time_str}. Total running time: " + running_time_str
+        )
+        self._print_result(trial, result)
+        print("")
+
+    def on_trial_complete(
+        self, iteration: int, trials: List[Trial], trial: Trial, **info
+    ):
+        if self.verbosity < self._start_end_verbosity:
+            return
+        curr_time_str, running_time_str = _get_time_str(self._start_time, time.time())
+        finished_iter = 0
+        if trial.last_result and TRAINING_ITERATION in trial.last_result:
+            finished_iter = trial.last_result[TRAINING_ITERATION]
+        print(
+            f"{self._addressing_tmpl.format(trial)} "
+            f"completed after {finished_iter} iterations "
+            f"at {curr_time_str}. Total running time: " + running_time_str
+        )
+        self._print_result(trial)
+        print("")
+
+    def on_checkpoint(
+        self,
+        iteration: int,
+        trials: List[Trial],
+        trial: Trial,
+        checkpoint: "_TrackedCheckpoint",
+        **info,
+    ):
+        if self._verbosity < self._intermediate_result_verbosity:
+            return
+        # don't think this is supposed to happen but just to be save.
+        saved_iter = "?"
```

Comment:
Can we set `saved_iter = 0` by default?

Reply:
Ah this is just moved from below - I'll leave it as is